contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,148 @@
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+pub struct DatabaseConfig<'a> {
+    pub url: &'a str,
+    pub pool_size: u32,
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the r2d2 connection pool methods.
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
+    from.get_slice("databases").and_then(|db_config| {
+        db_config
+            .iter()
+            .filter(|element|
+                match element.as_table().and_then(|table|
+                    Some(table["name"].as_str() == Some(name))
+                ) {
+                    Some(true) => true,
+                    _ => false,
+                }
+            )
+            .nth(0)
+            .ok_or(ConfigError::UnknownKey(format!("Could not find key: {}", name)))
+    })
+    .and_then(|connection_config| {
+        Ok(connection_config["url"].as_str().and_then(|url| {
+            Some(DatabaseConfig {
+                url,
+                pool_size: connection_config["pool_size"].as_integer().unwrap_or(from.workers as i64) as u32,
+                extras: connection_config.as_table(),
+            })
+        }))
+    })
+    .ok()
+    .and_then(|result| result)
+}
+
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Vec<&'a str> {
+    let db_config = config.get_slice("databases").unwrap();
+    db_config
+        .iter()
+        .map(|config| {
+            config.as_table().and_then(|table| table["name"].as_str())
+        })
+        .filter(|e| e.is_some())
+        .map(|e| e.unwrap())
+        .collect::<Vec<&'a str>>()
+}
+
+/// This trait represents a database connection that can be managed with
+/// R2D2. Implementing this on a connection type allows Rocket to automatically
+/// generate the code for a connection pool.
+pub trait Poolable: Send + Sized + 'static {
+    type Manager: ManageConnection<Connection=Self>;
+    type Error;
+
+    fn manager(config: &DatabaseConfig) -> Result<Self::Manager, Self::Error>;
+}
+
+#[cfg(feature = "diesel_sqlite")]
+impl Poolable for diesel::SqliteConnection {
+    type Manager = diesel::r2d2::ConnectionManager<diesel::SqliteConnection>;
+    type Error = DbError<io::Error>;
+
+    fn manager(config: &DatabaseConfig) -> Result<Self::Manager, Self::Error> {
+        Ok(diesel::r2d2::ConnectionManager::new(config.url))
```

Comment:
This is weird. What's the point of the `map_err`? It looks like this can never fail, is that right? If so, the error type should be `!`.

Reply:
I wasn't aware of the never type until now. Retrieving the `ConnectionManager` from Diesel never can fail, so it would make sense that its `Error` type is `!`.